src/Optimization/hiopAlgPrimalDecomp.cpp

```diff
@@ -1204,6 +1236,523 @@ void hiopAlgPrimalDecomposition::set_alpha_max(const double alp_max)
+        double aux=0.;
+
+        if(nc_<n_) {
+          x0->copy_from_indexes(*x_, *xc_idx_);
+        } else {
+          assert(nc_==n_);
+          x0->copyFromStarting(0, *x_);
+        }
+        for(int ri=0; ri<cont_idx.size(); ri++) {
+          aux = 0.;
+          int idx_temp = cont_idx[ri];
+
+          bret = master_prob_->eval_f_rterm(idx_temp, nc_, x0_vec, aux); // solving the recourse problem
+          if(!bret) {
+              //TODO
+          }
+          //rec_val += aux;
+          rval += aux;
+        }
+
+        // log_->printf(hovFcnEval, "recourse value: is %18.12e)\n", rec_val);
+        hiopVector* grad_aux = x0->alloc_clone();
+        grad_aux->setToZero(); 
+
+        for(int ri=0; ri<cont_idx.size(); ri++) {
+          int idx_temp = cont_idx[ri];
+          bret = master_prob_->eval_grad_rterm(idx_temp, nc_, x0_vec, *grad_aux);
+          if(!bret) {
+            //TODO
+          }
+          //grad_acc->axpy(1.0, *grad_aux);
+          grad_r->axpy(1.0, *grad_aux);
+        }
+        delete grad_aux;
+
+        // no need to set values for rec_prob anymore
+        //rec_prob[my_rank_]->set_value(rec_val);
+        //rec_prob[my_rank_]->set_grad(grad_acc_vec);
+        
+        //rec_prob[my_rank_]->post_send(2, rank_master, comm_world_);
+        // send signal that subproblem has been solved
+        rec_prob[my_rank_]->post_send_end_signal(2, rank_master, comm_world_);
+
+        // request the next subproblem index 
+        req_cont_idx->post_recv(1, rank_master, comm_world_);
+        while(cont_idx[0]!=-1) {//loop until end signal received
+          mpi_test_flag = req_cont_idx->test();
+          /* contigency starts at 0 
+           * sychronous implmentation of contingencist
+          */
+          if(mpi_test_flag) {
+            // log_->printf(hovIteration, "contingency index %d, rank %d)\n", cont_idx[0],my_rank_); 
+            for(int ri=0; ri<cont_idx.size(); ri++) {
+              cont_idx[ri] = req_cont_idx->value();
+            }
+            if(cont_idx[0]==-1) {
+              break;
+            }
+            
+            // accumulate locally so cannot set to zero
+            //rec_val = 0.;
+            //grad_acc->setToZero();
+
+            double aux=0.;
+            if(nc_<n_) {
+              x0->copy_from_indexes(*x_, *xc_idx_);
+            } else {
+              assert(nc_==n_);
+              x0->copyFromStarting(0, *x_);
+            }
+            for(int ri=0; ri<cont_idx.size(); ri++) {
+              aux = 0.;
+              int idx_temp = cont_idx[ri];
+              
+              bret = master_prob_->eval_f_rterm(idx_temp, nc_, x0_vec, aux); //need to add extra time here
+              if(!bret) {
+              //TODO
+              }
+              //rec_val += aux;
+              rval += aux;
+            }
+            // log_->printf(hovFcnEval, "recourse value: is %18.12e)\n", rec_val);
+            hiopVector* grad_aux = x0->alloc_clone();
```

Comment:
move `grad_aux` outside the loop to avoid costly, unnecessary allocs and deallocs